source/common/websocket/codec.cc

```diff
@@ -0,0 +1,218 @@
+  } else {
+    state_ = State::KFramePayload;
+  }
+}
+
+void Decoder::frameData(const uint8_t* mem, uint64_t length) { frame_.payload_->add(mem, length); }
+
+void Decoder::frameDataEnd() {
+  if (!output_) {
+    output_ = std::vector<Frame>();
+  }
+  output_.value().push_back(std::move(frame_));
+  frame_.final_fragment_ = false;
+  frame_.opcode_ = 0;
+  frame_.payload_length_ = 0;
+  frame_.payload_ = nullptr;
+  frame_.masking_key_ = absl::nullopt;
+}
+
+void Decoder::doDecodeMaskFlagAndLength(const uint8_t c, const uint8_t*& mem,
+                                        uint64_t& slice_index) {
+  frameMaskFlag(c);
+  if (length_ == 0x7e) {
+    length_of_extended_length_ = kPayloadLength16Bit;
+    length_ = 0;
+    state_ = State::KFrameHeaderExtendedLength;
+  } else if (length_ == 0x7f) {
+    length_of_extended_length_ = kPayloadLength64Bit;
+    length_ = 0;
+    state_ = State::KFrameHeaderExtendedLength;
+  } else if (masking_key_length_ > 0) {
+    state_ = State::KFrameHeaderMaskingKey;
+  } else {
+    frameDataStart();
+  }
+  slice_index++;
+  mem++;
+}
+
+void Decoder::doDecodeExtendedLength(const uint8_t c, const uint8_t*& mem, uint64_t& slice_index) {
+  if (length_of_extended_length_ == 1) {
+    length_ |= static_cast<uint64_t>(c);
+  } else {
+    length_ |= static_cast<uint64_t>(c) << 8 * (length_of_extended_length_ - 1);
+  }
+  length_of_extended_length_--;
+  if (length_of_extended_length_ == 0) {
+    if (masking_key_length_ > 0) {
+      state_ = State::KFrameHeaderMaskingKey;
+    } else {
+      frameDataStart();
+    }
+  }
+  slice_index++;
+  mem++;
+}
+
+void Decoder::doDecodeMaskingKey(const uint8_t c, const uint8_t*& mem, uint64_t& slice_index) {
+  if (masking_key_length_ == 1) {
+    masking_key_ |= static_cast<uint32_t>(c);
+  } else {
+    masking_key_ |= static_cast<uint32_t>(c) << 8 * (masking_key_length_ - 1);
+  }
+  masking_key_length_--;
+  if (masking_key_length_ == 0) {
+    frameMaskingKey();
+    frameDataStart();
+  }
+  slice_index++;
+  mem++;
+}
+
+void Decoder::doDecodePayload(const Buffer::RawSlice& slice, const uint8_t*& mem,
+                              uint64_t& slice_index) {
+  uint64_t remain_in_buffer = slice.len_ - slice_index;
+  if (remain_in_buffer <= length_) {
+    frameData(mem, remain_in_buffer);
+    mem += remain_in_buffer;
+    slice_index += remain_in_buffer;
+    length_ -= remain_in_buffer;
+  } else {
+    frameData(mem, length_);
+    mem += length_;
+    slice_index += length_;
+    length_ = 0;
+  }
+  if (length_ == 0) {
+    frameDataEnd();
+    state_ = State::KFrameHeaderFlagsAndOpcode;
+  }
+}
+
+absl::optional<std::vector<Frame>> Decoder::decode(Buffer::Instance& input) {
+  decoding_error_ = false;
+  inspect(input);
+  if (decoding_error_) {
+    return absl::nullopt;
+  }
+  input.drain(input.length());
+  return std::move(output_);
+}
+
+uint64_t Decoder::inspect(const Buffer::Instance& data) {
+  uint64_t frames_count_ = 0;
```

Comment:
nit: local variables should not end with `_`, that's reserved for class members.